lib/SILOptimizer/Analysis/DifferentiableActivityAnalysis.cpp

```diff
@@ -343,18 +344,25 @@ void DifferentiableActivityInfo::setUsefulThroughArrayInitialization(
       auto *ptai = dyn_cast<PointerToAddressInst>(use->getUser());
       assert(ptai && "Expected `pointer_to_address` user for uninitialized "
                      "array intrinsic");
-      // Propagate usefulness through array element addresses.
-      // - Find `store` and `copy_addr` instructions with array element
-      //   address destinations.
-      // - For each instruction, set destination (array element address) as
-      //   useful and propagate usefulness through source.
+      setUseful(ptai, dependentVariableIndex);
+      // Propagate usefulness through array element addresses:
+      // `pointer_to_address` and `index_addr` instructions.
+      //
+      // - Set all array element addresses as useful.
+      // - Find instructions with array element addresses as "result":
+      //   - `store` and `copy_addr` with array element address as destination.
+      //   - `apply` with array element address as an indirect result.
+      // - For each instruction, propagate usefulness through "arguments":
+      //   - `store` and `copy_addr`: propagate to source.
+      //   - `apply`: propagate to arguments.
       //
-      // Note: `propagateUseful(use->getUser(), ...)` is intentionally not used
+      // NOTE: `propagateUseful(use->getUser(), ...)` is intentionally not used
```

Comment:
Note: rather than pattern-matching yet another special case of "writing to an array literal element address" (`store`, `copy_addr`, and now `apply` indirect result) here in `DifferentiableActivityInfo::setUsefulThroughArrayInitialization`, it may be nice to call a blanket `propagateUseful(use->getUser(), ...)`, at risk of marking some "junk" values as useful:
* The `array.uninitialized_intrinsic` `RawPointer` result
* `integer_literal` operands to `index_addr` element addresses.

In practice, calling `propagateUseful(use->getUser(), ...)` is blocked by TF-1032, which I plan to investigate soon.